x/ibc/02-client/handler.go

```diff
@@ -10,34 +10,28 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/keeper"
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
-	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/07-tendermint/types"
-	localhosttypes "github.com/cosmos/cosmos-sdk/x/ibc/09-localhost/types"
 )
 
 // HandleMsgCreateClient defines the sdk.Handler for MsgCreateClient
 func HandleMsgCreateClient(ctx sdk.Context, k keeper.Keeper, msg exported.MsgCreateClient) (*sdk.Result, error) {
 	clientType := exported.ClientTypeFromString(msg.GetClientType())
 
-	var (
-		clientState     exported.ClientState
-		consensusHeight uint64
-	)
+	var consensusHeight uint64
 
-	switch clientType {
-	case exported.Tendermint:
-		tmMsg, ok := msg.(*ibctmtypes.MsgCreateClient)
-		if !ok {
-			return nil, sdkerrors.Wrapf(types.ErrInvalidClientType, "got %T, expected %T", msg, &ibctmtypes.MsgCreateClient{})
-		}
+	clientState := msg.InitializeClientState()
 
-		clientState = ibctmtypes.InitializeFromMsg(tmMsg)
-		consensusHeight = msg.GetConsensusState().GetHeight()
+	switch clientType {
 	case exported.Localhost:
-		// msg client id is always "localhost"
-		clientState = localhosttypes.NewClientState(ctx.ChainID(), ctx.BlockHeight())
+		// Localhost consensus height is chain's blockheight
 		consensusHeight = uint64(ctx.BlockHeight())
 	default:
-		return nil, sdkerrors.Wrapf(types.ErrInvalidClientType, "unsupported client type (%s)", msg.GetClientType())
+		consState := msg.GetConsensusState()
+		if consState == nil {
```

Comment:
I'm not sure that returning an error is good? I for eg the localhost client would return `nil` here. I'd just set the `consensusHeight  = 0 in this case`

Reply:
Went with fede's suggestion